README.md
# Tiles API

Add tiles API to QGIS Server
* Tiles Map Service API
* OGC Tiles API

## Tile Map Service API - TMS

The TMS API provides these URLs:
* `/tms/?`
  * to get information on the Tile Map Service based on WMTS Server project configuration
* `/tms/tilemaps/(?<tileMapId>[^/]+)/?`
  * to get information on a tile map based on WMTS Server project configuration
* `/tms/tilemaps/(?<tileMapId>[^/]+)/(?P<tilematrixid>\d+)/(?P<tilecolid>\d+)/(?P<tilerowid>\d+)\.(?P<extension>[^/?]+)/?`
  * to get tile map content based on extension
  * the available extension is png, jpg, jpeg and pbf
  * the image formats, png and jpg, is configured in the project
  * the vector tile format is based on QGIS version > 3.14 and only available for vector layers


CHANGELOG.md
# Changelog

## Unreleased

* Tile Map Server API - TMS



tests/conftest.py
import configparser
import glob
import logging
import os
import sys

import lxml.etree
import pytest

logging.basicConfig( stream=sys.stderr )
logging.disable(logging.NOTSET)

LOGGER = logging.getLogger('server')
LOGGER.setLevel(logging.DEBUG)

from typing import Any, Dict, Generator

from qgis.core import Qgis, QgsApplication, QgsFontUtils, QgsProject
from qgis.server import (
    QgsBufferServerRequest,
    QgsBufferServerResponse,
    QgsServer,
    QgsServerRequest,
)

qgis_application = None

def pytest_addoption(parser):
    parser.addoption("--qgis-plugins", metavar="PATH", help="Plugin path", default=None)


plugin_path = None


def pytest_configure(config):
    global plugin_path
    plugin_path = config.getoption('qgis_plugins')


def pytest_sessionstart(session):
    """ Start qgis application
    """
    global qgis_application
    os.environ['QT_QPA_PLATFORM'] = 'offscreen'

    # Define this in global environment
    #os.environ['QGIS_DISABLE_MESSAGE_HOOKS'] = 1
    #os.environ['QGIS_NO_OVERRIDE_IMPORT'] = 1
    qgis_application = QgsApplication([], False)
    qgis_application.initQgis()

    # Install logger hook
    install_logger_hook()


def pytest_sessionfinish(session, exitstatus):
    """ End qgis session
    """
    global qgis_application
    qgis_application.exitQgis()
    del qgis_application


NAMESPACES = {
    'xlink': "http://www.w3.org/1999/xlink",
    'wms': "http://www.opengis.net/wms",
    'wfs': "http://www.opengis.net/wfs",
    'wcs': "http://www.opengis.net/wcs",
    'ows': "http://www.opengis.net/ows/1.1",
    'gml': "http://www.opengis.net/gml",
    'xsi': "http://www.w3.org/2001/XMLSchema-instance"
}

class OWSResponse:

    def __init__(self, resp: QgsBufferServerResponse) -> None:
        self._resp = resp
        self._xml = None

    @property
    def xml(self) -> 'xml':
        if self._xml is None and self._resp.headers().get('Content-Type','').find('text/xml')==0:
            self._xml = lxml.etree.fromstring(self.content.decode('utf-8'))
        return self._xml

    @property
    def content(self) -> bytes:
        return bytes(self._resp.body())

    @property
    def status_code(self) -> int:
        return self._resp.statusCode()

    @property
    def headers(self) -> Dict[str,str]:
        return self._resp.headers()

    def xpath(self, path: str) -> lxml.etree.Element:
        assert self.xml is not None
        return self.xml.xpath(path, namespaces=NAMESPACES)

    def xpath_text(self, path: str) -> str:
        assert self.xml is not None
        return ' '.join(e.text for e in self.xpath(path))


@pytest.fixture(scope='session')
def client(request):
    """ Return a qgis server instance
    """
    class _Client:

        def __init__(self) -> None:
            self.fontFamily = QgsFontUtils.standardTestFontFamily()
            QgsFontUtils.loadStandardTestFonts(['All'])

            # Activate debug headers
            os.environ['QGIS_WMTS_CACHE_DEBUG_HEADERS'] = 'true'

            self.datapath = request.config.rootdir.join('data')
            self.server = QgsServer()

            # Load plugins
            load_plugins(self.server.serverInterface())

        def getplugin(self, name) -> Any:
            """ retourne l'instance du plugin
            """
            return server_plugins.get(name)

        def getprojectpath(self, name: str) -> str:
            return self.datapath.join(name)

        def request(self, method, query:str, project: str=None, data: bytes=None,
                    headers: dict={}) -> OWSResponse:
            """ Return server response from query
            """
            request  = QgsBufferServerRequest(query, method, headers, data)
            response = QgsBufferServerResponse()
            if project is not None and not os.path.isabs(project):
                projectpath = self.datapath.join(project)
                qgsproject  = QgsProject()
                if not qgsproject.read(projectpath.strpath):
                    raise ValueError(f"Error reading project '{projectpath.strpath}'")
            else:
                qgsproject = None
            self.server.handleRequest(request, response, project=qgsproject)
            return OWSResponse(response)

        def get(self, *args, **kwargs) -> OWSResponse:
            return self.request(QgsServerRequest.GetMethod, *args, **kwargs)

        def delete(self, *args, **kwargs) -> OWSResponse:
            return self.request(QgsServerRequest.DeleteMethod, *args, **kwargs)

    return _Client()


##
## Plugins
##

def checkQgisVersion(minver: str, maxver: str) -> bool:

    def to_int(ver):
        major, *ver = ver.split('.')
        major = int(major)
        minor = int(ver[0]) if len(ver) > 0 else 0
        rev   = int(ver[1]) if len(ver) > 1 else 0
        if minor >= 99:
            minor = rev = 0
            major += 1
        if rev > 99:
            rev = 99
        return int("{:d}{:02d}{:02d}".format(major,minor,rev))


    version = to_int(Qgis.QGIS_VERSION.split('-')[0])
    minver  = to_int(minver) if minver else version
    maxver  = to_int(maxver) if maxver else version
    return minver <= version <= maxver


def find_plugins(pluginpath: str) -> Generator[str,None,None]:
    """ Load plugins
    """
    for plugin in glob.glob(os.path.join(plugin_path + "/*")):
        if not os.path.exists(os.path.join(plugin, '__init__.py')):
            continue

        metadatafile = os.path.join(plugin, 'metadata.txt')
        if not os.path.exists(metadatafile):
            continue

        cp = configparser.ConfigParser()
        try:
            with open(metadatafile, mode='rt') as f:
                cp.read_file(f)
            if not cp['general'].getboolean('server'):
                logging.critical("%s is not a server plugin", plugin)
                continue

            minver = cp['general'].get('qgisMinimumVersion')
            maxver = cp['general'].get('qgisMaximumVersion')
        except Exception as exc:
            LOGGER.critical("Error reading plugin metadata '%s': %s",metadatafile,exc)
            continue

        if not checkQgisVersion(minver,maxver):
            LOGGER.critical(("Unsupported version for %s:"
                "\n MinimumVersion: %s"
                "\n MaximumVersion: %s"
                "\n Qgis version: %s"
                "\n Discarding") % (plugin,minver,maxver,
                    Qgis.QGIS_VERSION.split('-')[0]))
            continue

        yield os.path.basename(plugin)


server_plugins = {}


def load_plugins(serverIface: 'QgsServerInterface') -> None:
    """ Start all plugins
    """
    if not plugin_path:
        return

    LOGGER.info("Initializing plugins from %s", plugin_path)
    sys.path.append(plugin_path)

    for plugin in find_plugins(plugin_path):
        try:
            __import__(plugin)

            package = sys.modules[plugin]

            # Initialize the plugin
            server_plugins[plugin] = package.serverClassFactory(serverIface)
            LOGGER.info("Loaded plugin %s",plugin)
        except:
            LOGGER.error("Error loading plugin %s",plugin)
            raise


#
# Logger hook
#

def install_logger_hook( verbose: bool=False ) -> None:
    """ Install message log hook
    """
    from qgis.core import Qgis, QgsApplication, QgsMessageLog

    # Add a hook to qgis  message log
    def writelogmessage(message, tag, level):
        arg = '{}: {}'.format( tag, message )
        if level == Qgis.Warning:
            LOGGER.warning(arg)
        elif level == Qgis.Critical:
            LOGGER.error(arg)
        elif verbose:
            # Qgis is somehow very noisy
            # log only if verbose is set
            LOGGER.info(arg)

    messageLog = QgsApplication.messageLog()
    messageLog.messageReceived.connect( writelogmessage )


tests/run-tests.sh
#!/bin/bash

set -e

# Add /.local to path
export PATH=$PATH:/.local/bin

echo "Installing required packages..."
pip3 install -q -U --prefer-binary --user -r tests/requirements.txt

# Disable qDebug stuff that bloats test outputs
export QT_LOGGING_RULES="*.debug=false;*.warning=false"

# Disable python hooks/overrides
export QGIS_DISABLE_MESSAGE_HOOKS=1
export QGIS_NO_OVERRIDE_IMPORT=1

cd tests

py.test -v --qgis-plugins=/src $@



tests/test_tmsapi.py
import json
import logging
import os

from shutil import rmtree

import lxml.etree

from qgis.core import QgsProject, Qgis

LOGGER = logging.getLogger('server')


def test_tmsapi_no_project(client):
    """ Test Project file error
    """
    plugin = client.getplugin('tilesForServer')
    assert plugin is not None

    # TMS API request
    qs = "/tms?"
    rv = client.get(qs)
    assert rv.status_code == 500
    assert rv.headers.get('Content-Type',"").startswith('application/json')

    json_content = json.loads(rv.content)
    assert 'httpcode' in json_content
    assert json_content['httpcode'] == 500
    assert 'status' in json_content
    assert json_content['status'] == 'error'
    assert 'error' in json_content
    assert 'message' in json_content['error']

    # Get undefined project
    project = QgsProject()
    project.setFileName(client.getprojectpath("undefined.qgs").strpath)

    # TMS API request
    qs = "/tms?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 500
    assert rv.headers.get('Content-Type',"").startswith('application/json')

def test_tmsapi_landingpage(client):
    """ Test the TMS API -Landing page
        /tms?
    """
    plugin = client.getplugin('tilesForServer')
    assert plugin is not None

    # Get project
    project = QgsProject()
    project.setFileName(client.getprojectpath("france_parts.qgs").strpath)

    # TMS API requests
    qs = "/tms?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 200
    assert rv.headers.get('Content-Type',"").startswith('application/json')

    json_content = json.loads(rv.content)
    assert 'title' in json_content

    assert 'abstract' in json_content

    assert 'tileMaps' in json_content
    assert len(json_content['tileMaps']) == 1
    tile_map = json_content['tileMaps'][0]
    assert 'id' in tile_map
    assert tile_map['id'] == 'france_parts'
    assert 'title' in tile_map
    assert tile_map['title'] == 'france_parts'
    assert 'abstract' in tile_map
    assert 'formats' in tile_map
    if Qgis.QGIS_VERSION_INT >= 31400:
        assert len(tile_map['formats']) == 2
    else:
        assert len(tile_map['formats']) == 1
    assert 'png' in tile_map['formats']
    if Qgis.QGIS_VERSION_INT >= 31400:
        assert 'pbf' in tile_map['formats']
    assert 'links' in tile_map
    assert len(tile_map['links']) == 1

    assert 'source_id' not in tile_map
    assert 'source_type' not in tile_map

    assert 'links' in json_content
    assert len(json_content['links']) == 0


    # Project with WMTS but not in EPSG:3857
    project.setFileName(client.getprojectpath("france_parts_grid_4326.qgs").strpath)
    qs = "/tms?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 200
    assert rv.headers.get('Content-Type',"").startswith('application/json')

    json_content = json.loads(rv.content)
    assert 'tileMaps' in json_content
    assert len(json_content['tileMaps']) == 0

    # Project without WMTS
    project.setFileName(client.getprojectpath("france_parts_no_wmts.qgs").strpath)
    qs = "/tms?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 200
    assert rv.headers.get('Content-Type',"").startswith('application/json')

    json_content = json.loads(rv.content)
    assert 'tileMaps' in json_content
    assert len(json_content['tileMaps']) == 0

def test_tmsapi_tilemapinfo(client):
    """ Test the TMS API -Landing page
        /tms/tilemaps/{tilemapid}?
    """
    plugin = client.getplugin('tilesForServer')
    assert plugin is not None

    # Get project
    project = QgsProject()
    project.setFileName(client.getprojectpath("france_parts.qgs").strpath)

    # TMS API requests
    qs = "/tms/tilemaps/france_parts?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 200
    assert rv.headers.get('Content-Type',"").startswith('application/json')

    json_content = json.loads(rv.content)
    assert 'id' in json_content
    assert json_content['id'] == 'france_parts'
    assert 'title' in json_content
    assert json_content['title'] == 'france_parts'
    assert 'abstract' in json_content
    assert 'bbox' in json_content
    assert 'formats' in json_content
    if Qgis.QGIS_VERSION_INT >= 31400:
        assert len(json_content['formats']) == 2
    else:
        assert len(json_content['formats']) == 1
    assert 'extension' in json_content['formats'][0]
    assert json_content['formats'][0]['extension'] == 'png'
    assert 'mimetype' in json_content['formats'][0]
    assert json_content['formats'][0]['mimetype'] == 'image/png'
    if Qgis.QGIS_VERSION_INT >= 31400:
        assert 'extension' in json_content['formats'][1]
        assert json_content['formats'][1]['extension'] == 'pbf'
        assert 'mimetype' in json_content['formats'][1]
        assert json_content['formats'][1]['mimetype'] == 'application/x-protobuf'

    # TMS API request - tilemapid unknwon
    qs = "/tms/tilemaps/unknwon?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 404
    assert rv.headers.get('Content-Type',"").startswith('application/json')

    json_content = json.loads(rv.content)
    assert 'httpcode' in json_content
    assert json_content['httpcode'] == 404
    assert 'status' in json_content
    assert json_content['status'] == 'error'
    assert 'error' in json_content
    assert 'message' in json_content['error']

def test_tmsapi_tilemapcontent(client):
    """ Test the TMS API -Landing page
        /tms/tilemaps/{tilemapid}/{tilematrixid}/{tilecolid}/{tilerowid}?
    """
    plugin = client.getplugin('tilesForServer')
    assert plugin is not None

    # Get project
    project = QgsProject()
    project.setFileName(client.getprojectpath("france_parts.qgs").strpath)

    # TMS API requests
    qs = "/tms/tilemaps/france_parts/0/0/0.png?MAP=%s" % project.fileName()
    rv = client.get(qs)
    assert rv.status_code == 200
    assert rv.headers.get('Content-Type',"").startswith('image/png')
    assert len(rv.content) > 0

    if Qgis.QGIS_VERSION_INT >= 31400:
        qs = "/tms/tilemaps/france_parts/0/0/0.pbf?MAP=%s" % project.fileName()
        rv = client.get(qs)
        assert rv.status_code == 200
        assert rv.headers.get('Content-Type',"").startswith('application/x-protobuf')
        assert len(rv.content) > 0

    # TMS API request - No project
    qs = "/tms/tilemaps/france_parts/0/0/0.png"
    rv = client.get(qs)
    assert rv.status_code == 500
    assert rv.headers.get('Content-Type',"").startswith('application/json')


tilesForServer/tmsapi.py
import os
import tempfile

from pathlib import Path

from qgis.server import (
    QgsServerProjectUtils,
    QgsServerOgcApi,
    QgsBufferServerRequest,
    QgsServerRequest,
)
from qgis.core import (
    Qgis,
    QgsMessageLog,
    QgsCoordinateTransform,
    QgsTileXYZ,
    QgsTileMatrix,
    #QgsVectorTileMVTEncoder, #define SIP_NO_FILE
    QgsVectorTileWriter,
    QgsFeedback,
    QgsMapLayer,
    QgsDataSourceUri,
)
from qgis.PyQt.QtCore import QUrl

from .apiutils import HTTPError, RequestHandler, register_api_handlers


#
# WMTS API Handlers
#

class ProjectParser:

    def tile_project_info(self):
        project = self._context.project()

        # The project as tiles source
        if not project.readBoolEntry("WMTSLayers", "Project")[0]:
            return

        # project identifier
        root_name = QgsServerProjectUtils.wmsRootName(project)
        if not root_name:
            root_name = project.title()
        if not root_name:
            return

        # available formats
        formats = []
        if project.readBoolEntry("WMTSPngLayers", "Project")[0]:
            formats.append('png')
        if project.readBoolEntry("WMTSJpegLayers", "Project")[0]:
            formats.append('jpg')

        if self.support_pbf():
            for layer in self._project.mapLayers().values():
                if layer.type() == QgsMapLayer.VectorLayer:
                    formats.append('pbf')
                    break

        if not formats:
            return

        # tileMap info
        yield {
                'source_id': root_name,
                'source_type': 'project',
                'id': root_name,
                'title': project.title(),
                'abstract': project.title(),
                'formats': formats
              }

    def tile_groups_info(self):
        project = self._context.project()

        # Groups as tiles source
        g_names = project.readListEntry("WMTSLayers", "Group")[0]
        if not g_names:
            return

        tree_root = project.layerTreeRoot()

        png_g_names = project.readListEntry("WMTSPngLayers", "Group")[0]
        jpg_g_names = project.readListEntry("WMTSJpegLayers", "Group")[0]

        for g_name in g_names:
            tree_group = tree_root.findGroup(g_name)
            if not tree_group:
                continue

            # Group identifier and other infos
            g_id = tree_group.customProperty("wmsShortName")
            if not g_id:
                g_id = g_name
            g_title = tree_group.customProperty("wmsTitle")
            if not g_title:
                g_title = g_name
            g_abstract = tree_group.customProperty("wmsAbstract")

            # available formats
            g_formats = []
            if g_name in png_g_names:
                g_formats.append('png')
            if g_name in jpg_g_names:
                g_formats.append('jpg')

            if self.support_pbf():
                for tree_layer in tree_group.findLayers():
                    layer = tree_layer.layer
                    if not layer:
                        continue
                    if layer.type() == QgsMapLayer.VectorLayer:
                        g_formats.append('pbf')
                        break

            if not g_formats:
                continue

            # tileMap info
            yield {
                    'source_id': g_name,
                    'source_type': 'group',
                    'id': g_id,
                    'title': g_title,
                    'abstract': g_abstract,
                    'formats': g_formats,
                  }

    def tile_layers_info(self):
        project = self._context.project()

        # Layers as tiles source
        layer_ids = project.readListEntry("WMTSLayers", "Layer")[0]
        if not layer_ids:
            return

        png_layer_ids = project.readListEntry("WMTSPngLayers", "Layer")[0]
        jpg_layer_ids = project.readListEntry("WMTSJpegLayers", "Layer")[0]

        for layer_id in layer_ids:
            layer = project.mapLayer(layer_id)
            if not layer:
                continue

            # Layer identifier and informations
            l_id = layer.shortName()
            if not l_id:
                l_id = layer.name()
            l_title = layer.title()
            if not l_title:
                l_title = layer.name()

            # available formats
            l_formats = []
            if layer_id in png_layer_ids:
                l_formats.append('png')
            if layer_id in jpg_layer_ids:
                l_formats.append('jpg')

            if self.support_pbf() and \
               layer.type() == QgsMapLayer.VectorLayer:
                l_formats.append('pbf')

            if not l_formats:
                continue

            # tileMap info
            yield {
                    'source_id': layer.id(),
                    'source_type': 'layer',
                    'id': l_id,
                    'title': l_title,
                    'abstract': layer.abstract(),
                    'formats': l_formats,
                  }

    def tile_maps_info(self):
        # The project as tiles source
        for info in self.tile_project_info():
            yield info

        # Groups as tiles source
        for info in self.tile_groups_info():
            yield info

        # Layers as tiles source
        for info in self.tile_layers_info():
            yield info

    def get_complete_tilemap_info(self, tilemapid):
        for info in self.tile_maps_info():
            if tilemapid != info['id']:
                continue

            extra = {**info}
            source_type = extra.pop('source_type')
            source_id = extra.pop('source_id')

            extra['bbox'] = self.get_tilemap_bbox(source_type, source_id)
            extra['formats'] = [{
                'extension': ext,
                'mimetype': self.mimetypeFromExtension(ext)
            } for ext in extra['formats']]
            return extra
        return None

    def get_tilemap_bbox(self, source_type, source_id):
        bbox = None
        if source_type == 'project':
            bbox = self.get_project_bbox()
        elif source_type == 'group':
            bbox = self.get_group_bbox(source_id)
        elif source_type == 'Layer':
            bbox = self.get_layer_bbox(source_id)
        return bbox

    def get_project_bbox(self):
        project = self._context.project()
        crs_dest = QgsCoordinateReferenceSystem("EPSG:3857")
        xform_context = project.transformContext()

        proj_rect = QgsServerProjectUtils.wmsExtent(project)
        xform = QgsCoordinateTransform(project.crs(), crs_dest, xform_context)
        proj_rect = xform.transform(proj_rect)

        return [
            proj_rect.xMinimum(), proj_rect.yMinimum(),
            proj_rect.xMaximum(), proj_rect.yMaximum()
        ]

    def get_group_bbox(self, group_name):
        project = self._context.project()
        crs_dest = QgsCoordinateReferenceSystem("EPSG:3857")
        xform_context = project.transformContext()

        group_rect = None
        tree_group = tree_root.findGroup(group_name)
        for tree_layer in tree_group.findLayers():
            layer = tree_layer.layer
            if not layer:
                continue

            xform = QgsCoordinateTransform(layer.crs(), crs_dest, xform_context)
            if not group_rect:
                group_rect = xform.transform(layer.extent())
            else:
                group_rect.combineExtentWith(xform.transform(layer.extent()))

        if not group_rect:
            return group_rect

        return [
            group_rect.xMinimum(), group_rect.yMinimum(),
            group_rect.xMaximum(), group_rect.yMaximum()
        ]

    def get_layer_bbox(self, layer_id):
        project = self._context.project()
        crs_dest = QgsCoordinateReferenceSystem("EPSG:3857")
        xform_context = project.transformContext()

        layer = project.mapLayer(layer_id)
        if not layer:
            return None

        xform = QgsCoordinateTransform(layer.crs(), crs_dest, xform_context)
        layer_rect = xform.transform(layer.extent())

        if not layer_rect:
            return layer_rect

        return [
            layer_rect.xMinimum(), layer_rect.yMinimum(),
            layer_rect.xMaximum(), layer_rect.yMaximum()
        ]

    def tilemap_vectorlayers(self, tilemapid):
        for info in self.tile_maps_info():
            if tilemapid != info['id']:
                continue

            source_type = info.get('source_type')
            source_id = info.get('source_id')
            if source_type == 'project':
                for layer in self._project.mapLayers().values():
                    if layer.type() == QgsMapLayer.VectorLayer:
                        yield layer
            elif source_type == 'group':
                tree_root = self._project.layerTreeRoot()
                tree_group = tree_root.findGroup(source_id)
                if not tree_group:
                    return
                for tree_layer in tree_group.findLayers():
                    layer = tree_layer.layer
                    if not layer:
                        continue
                    if layer.type() == QgsMapLayer.VectorLayer:
                        yield layer
            elif source_type == 'layer':
                layer = self._project.mapLayer(source_id)
                if not layer:
                    return
                if layer.type() == QgsMapLayer.VectorLayer:
                    yield layer

    def mimetypeFromExtension(self, extension):
        if extension == 'png':
            return 'image/png'
        elif extension in ('jpg', 'jpeg'):
            return 'image/jpeg'

        if self.support_pbf and extension == 'pbf':
            return 'application/x-protobuf'

        return ''

    def support_pbf(self):
        return Qgis.QGIS_VERSION_INT >= 31400


class LandingPage(RequestHandler, ProjectParser):
    """ Project tile map listing handler
    """
    def get(self) -> None:
        project = self._context.project()

        grids = project.readListEntry("WMTSGrids", "CRS")[0]

        # tileMaps generator
        def links():
            # the tms API is available only for EPSG:3857 grid
            if 'EPSG:3857' not in grids:
                return
            # tile maps
            for info in self.tile_maps_info():
                tile_map_id = info['id']
                extra = {**info}
                del extra['source_id']
                del extra['source_type']
                extra['links'] = [{
                        'href': self.href(f"/tilemaps/{tile_map_id})", QgsServerOgcApi.contentTypeToExtension(QgsServerOgcApi.JSON)),
                        "rel": QgsServerOgcApi.relToString(QgsServerOgcApi.item),
                        "type": QgsServerOgcApi.mimeType(QgsServerOgcApi.JSON),
                        "title": "Cache collection",
                    }]
                yield extra

        data = {
            'title': QgsServerProjectUtils.owsServiceTitle(project),
            'abstract': QgsServerProjectUtils.owsServiceAbstract(project),
            'tileMaps': list(links()),
            'links': [],  # self.links(context)
        }
        self.write(data)

class TileMapInfo(RequestHandler, ProjectParser):
    """ Tile map information handler
    """
    def get(self, tilemapid):
        QgsMessageLog.logMessage('Tile map id: %s' % tilemapid, "tilesApi", Qgis.Warning)
        info = self.get_complete_tilemap_info(tilemapid)
        if not info :
            raise HTTPError(404, 'Tile map not found')
        self.write(info)


class TileMapContent(RequestHandler, ProjectParser):
    """ Tile map content handler
    """
    def initialize(self, srv_iface, **kwargs ) -> None:
        """ override
        """
        super().initialize(**kwargs)
        self._srv_iface = srv_iface

    def get(self, tilemapid, tilematrixid, tilecolid, tilerowid, extension):
        project = self._context.project()

        mimetype = self.mimetypeFromExtension(extension)
        if not mimetype:
            raise HTTPError(404, 'Extension unknown')

        if self.support_pbf and extension == 'pbf':
            tile = QgsTileXYZ(int(tilecolid), int(tilerowid), int(tilematrixid))

            # QgsVectorTileMVTEncoder is not defined in SIP
            # we need to use QgsVectorTileWriter to create the VectorTile
            # encoder = QgsVectorTileMVTEncoder(tile)

            tilematrix = QgsTileMatrix.fromWebMercator(tile.zoomLevel())
            tile_extent = tilematrix.tileExtent(tile);

            xform_context = self._project.transformContext()
            # encoder.setTransformContext(xform_context)

            # feedback = QgsFeedback()
            layers = []
            for layer in self.tilemap_vectorlayers(tilemapid):
                #encoder.addLayer(layer, feedback, '', layer.name())
                layers.append(QgsVectorTileWriter.Layer(layer))

            #tileData = encoder.encode()
            vt_writer = QgsVectorTileWriter()
            vt_writer.setExtent(tile_extent)
            vt_writer.setMaxZoom(tile.zoomLevel())
            vt_writer.setMinZoom(tile.zoomLevel())
            vt_writer.setLayers(layers)

            tmp_dir = tempfile.gettempdir()
            ds = QgsDataSourceUri()
            ds.setParam("type", "xyz" )
            ds.setParam("url", QUrl.fromLocalFile(tmp_dir).toString() + '/{z}-{x}-{y}.pbf' )

            vt_writer.setDestinationUri(bytes(ds.encodedUri()).decode())
            if not vt_writer.writeTiles():
                raise HTTPError(500, vt_writer.errorMessage())

            pbf_path = tmp_dir+'/{z}-{x}-{y}.pbf'.format(**{
                'z': tilematrixid,
                'x': tilecolid,
                'y': tilerowid,
            })
            if not os.path.exists(pbf_path):
                raise HTTPError(500, 'Vector Tile not generated')

            self.set_header('Content-Type', mimetype)
            #self._response.write(tileData)
            with open(pbf_path, 'rb+') as pbf:
                self._response.write(pbf.read())
            os.remove(pbf_path)
            self.finish()
        else:
            parameters = {
                "MAP": project.fileName(),
                "SERVICE": "WMTS",
                "VERSION": "1.0.0",
                "REQUEST": "GetTile",
                "LAYER": tilemapid,
                "STYLE": "",
                "TILEMATRIXSET": "EPSG:3857",
                "TILEMATRIX": tilematrixid,
                "TILEROW": tilerowid,
                "TILECOL": tilecolid,
                "FORMAT": mimetype,
            }
            qs = "?" + "&".join("%s=%s" % item for item in parameters.items())
            req = QgsBufferServerRequest(qs, QgsServerRequest.GetMethod, {}, None)
            service = self._srv_iface.serviceRegistry().getService('WMTS', '1.0.0')
            service.executeRequest(req, self._response, project );
            self._finished = True


def init_tms_api(serverIface) -> None:
    """ Initialize the Tile Map Server API
    """
    kwargs = dict(srv_iface = serverIface)

    tilemapid = r"tilemaps/(?P<tilemapid>[^/?]+)"

    handlers = [
        (rf"/{tilemapid}/(?P<tilematrixid>\d+)/(?P<tilecolid>\d+)/(?P<tilerowid>\d+)\.(?P<extension>[^/?]+)/?", TileMapContent,  kwargs),
        (rf"/{tilemapid}/?", TileMapInfo,  kwargs),
        (rf"/?", LandingPage, kwargs),
    ]

    register_api_handlers(serverIface, '/tms', 'TileMapService', handlers)
